### ISSUE TITLE
Adding support fo conditions in RASA export

### ISSUE THREAD

**Issue by Gauhar** [Botmock/sdk#561]
# Problem

In Botmock we offer a rasa exporter that takes your Botmock project and creates multiple files. 

```jsx
├── domain.yml
├── data/
│   ├── nlu.yml
│   ├── stories.yml

```

Technically we create markdown files for rasa 1.0 but PR to shift to `yml` file structure for rasa 2.0 was already created.

The problem is our current exporter can't handle paths with conditions. To add support of conditions there are many things we need to add such as 

1st: Some way to alter story based on conditions, like if the name is Rahul follow this story else follow that story.

2nd: Some way to decide what should be the current reply.

There are some problems with adding conditions support in rasa. 

In Watson we can directly add conditions on intent, in Dialogflow we can use context to handle the correct flow of conversation. But in Rasa, we can't trigger particular intent based on any condition, also we can't use context in Rasa.

# Solution

In Rasa, we have an actions folder that contains [actions.py](http://actions.py) file where we can write our own custom logic in the response to any particular intent. 

This is very much similar to our Dialogflow webhook where the intent is triggered in Dialogflow but a response is evaluated and came from the webhook.

There will be other things we need to change in rasa exporter other than adding conditions in [actions.py](http://actions.py) file, like changing stories files to replace all replies that have conditions on intent with actions.

One problem was that we can handle the responsibility of an intent-based on user input using [actions.py](http://actions.py) file but how are we going to put rasa on the correct path. 

In dialogflow we use context, i.e dialogflow also gave a context of the next response, and only intent with that context will be triggered. But in rasa, we don't have context.

One thing we can use to solve this issue was to use the slot_was_set condition in rasa as a context checker. 

Let us take an example.

Suppose we want to export the following project on rasa.

<img width="708" alt="Screenshot 2021-01-27 at 9 34 36 PM" src="https://user-images.githubusercontent.com/31892209/106018383-76191800-60e7-11eb-9cfd-fe5a769b260f.png">

According to our plan, there will be 2 paths in stories. There are also two connectors from a message block, both contain the same intent but different conditions. Here we'll use [actions.py](http://actions.py) file to evaluate and give an output of name intent.

But after that, both paths have the same intent but only once can be executed based on the previous context. My plan is to pass a context that will set the path of the next intent. Just like in this story.yml file

```jsx
version: "2.0"

stories:

- story: get name rahul
  steps:
  - intent: greet
  - action: utter_greet
  - intent: name
  - action: action_say_name
  - slot_was_set:
    - context: "rahulPath"
  - intent: test
  -  action: utter_rahul

- story: get name john
  steps:
  - intent: greet
  - action: utter_greet
  - intent: name
  - action: action_say_name
  - slot_was_set:
    - context: "johnPath"
  - intent: test
  -  action: utter_john
```

Once the user says his name, the `action_say_name` will be triggered, evaluate the condition and give a proper response and also one of the contexts (like `rahulPath` or `johnPath` in this example). Till name intent, both stories are the same but from here only one will be executed based on slot value of context. 

If the user said I am John, then the webhook will reply "Hello John" and also set the value of context to `johnPath` now if the user says the text, only  `utter_john` action will be triggered because of the story structure.

# What needed to be done?

1. First we need to change all the files, mostly stories.yml to call action whenever we see a condition in botmock intent or any of child of that intent. 

    This is going to be very much similar to dialogflow and 90% of work is already done and the remaining can be done with small modifications in how we generate stories files.

2. Second we need to create condition.json, block.json, and [actions.py](http://actions.py) file. Code to create condition.json and block.json can extract from dialogflow with little effort. actions.py file can be created in the same way as we create function.js i.e by using templates.
3. In Dialogflow we use the botmockUtils module which contains different functions to evaluate conditions, handle responses, etc. We can't use the same module in rasa and we need to create the same botmockUtils module in python. It's also good to create from scratch because rasa didn't need many of those functions we created in botmockUtils for dialogflow.

    This whole process should not take 5-7 working days including all modifications in rasa exporter to generate additional files, change existing stories, domain, NLU file, and writing botmockUtils.
